lib/puppet/configurer.rb

```diff
@@ -71,13 +73,18 @@ def retrieve_catalog(query_options)
 
         if result
           @cached_catalog_status = 'on_failure'
+
+          if result.environment == @environment
+            Puppet.info "Using cached catalog from environment '#{result.environment}'"
+          else
+            Puppet.err "Not using cached catalog because its environment '#{result.environment}' does not match '#{@environment}'"
+            result = nil
```

Comment:
Some discussion about whether we should always fail the run here. If the agent is configured in an environment `production`, and the ENC normally assigns the node to a different one `dev`, then the next time the agent runs, if the node request fails, then the agent's current environment will never match the cached catalog's environment, and we'd never apply the cached catalog.